rust/src/sans_io/read.rs

```diff
@@ -0,0 +1,1017 @@
+                            }
+                            let opcode_len_buf =
+                                decompress!(9, &mut state.compressed_remaining, decompressor);
+                            let opcode = opcode_len_buf[0];
+                            let len = match len_as_usize(u64::from_le_bytes(
+                                opcode_len_buf[1..9].try_into().unwrap(),
+                            )) {
+                                Ok(n) => n,
+                                Err(err) => return Some(Err(err)),
+                            };
+                            let _ =
+                                decompress!(9 + len, &mut state.compressed_remaining, decompressor);
+                            self.uncompressed_content.mark_read(9);
+                            let (start, end) = (
+                                self.uncompressed_content.start,
+                                self.uncompressed_content.start + len,
+                            );
+                            self.uncompressed_content.mark_read(len);
+                            let data = &self.uncompressed_content.data[start..end];
+                            return Some(Ok(ReadAction::GetRecord { data, opcode }));
+                        }
+                    }
+                }
+                PaddingAfterChunk => {
+                    {
+                        let state = self
+                            .chunk_state
+                            .as_mut()
+                            .expect("chunk state should be set");
+                        let _ = consume!(self.file_data, state.padding_after_compressed_data);
+                        if let Some(mut decompressor) = state.decompressor.take() {
+                            if let Err(err) = decompressor.reset() {
+                                return Some(Err(err));
+                            }
+                            self.decompressors
+                                .insert(decompressor.name().into(), decompressor);
+                        }
+                    }
+                    self.chunk_state = None;
+                    self.currently_reading = FileRecord;
+                }
+            }
+        }
+    }
+}
+
+// impl Default for LinearReader {
+//     fn default() -> Self {
+//         Self::new()
+//     }
+// }
+
+/// Encapsulates the action the user should take next when reading an MCAP file.
+///
+pub enum ReadAction<'a> {
+    /// The reader needs more data to continue - call [`LinearReader::insert`] to load more data.
+    /// the value provided here is a hint for how much data to insert.
+    NeedMore(usize),
+    /// Read a record out of the MCAP file. Use [`parse_record`] to parse the record.
+    GetRecord { data: &'a [u8], opcode: u8 },
+}
+
+fn len_as_usize(len: u64) -> McapResult<usize> {
+    len.try_into().map_err(|_| McapError::TooLong(len))
+}
+
+fn clamp_to_usize(len: u64) -> usize {
+    len.try_into().unwrap_or(usize::MAX)
+}
```

Comment:
You should just be able to do `as usize` here right?

It probably doesn't matter here but `try_into` + `MAX` generates more assembly: https://play.rust-lang.org/?version=stable&mode=release&edition=2021&gist=e1784ebf677852e59782a0755b879292

Reply:
`as usize` will truncate the MSB rather than saturate, which is what I want here.

Reply:
Ahh of course 🤦🏻